platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,86 @@
+# Lifecycle management for the State
+
+## Summary
+
+This proposal describes a possible implementation of state lifecycle management that could work equally well for both the Consensus Node and the Block Node.
+
+| Metadata           | Entities                                   |
+|--------------------|--------------------------------------------|
+| Designers          | [@imalygin](https://github.com/imalygin) |
+| Functional Impacts | Consensus Node, Block Node                 |
+
+## Purpose and Context
+
+There is an initiative to create self-contained State modules (`swirlds-state-api` and `swirlds-state-impl`) that would have minimal dependencies, and specifically, they should not depend on `swirlds-platform-core`. Part of this initiative is the implementation of a state lifecycle management mechanism that could be used by both the Consensus Node and the Block Node.
+
+### Requirements
+
+- Add an interface representing state lifecycle management and its implementation.
+- Update the current State API to be coherent with this State Lifecycle API.
+- Implement this set of APIs and migrate the existing code to use it.
+
+## Design & Architecture
+
+### Java classes
+
+- `com.swirlds.state.State` has the following lifecycle-related methods:
+  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
+  - `void computeHash()`: computes the hash of the immutable state.
+  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
+  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.
+
+These methods correspond to the state lifecycle as follows:
+
+[![State lifecycle](state-lifecycle.svg)](state-lifecycle.svg)
+
+Note that the `loadSnapshot` method doesn't really belong to `com.swirlds.state.State`, as it's a method of the class, and in some cases, we may need to load the snapshot before we have an instance of the state. 
+This method should be moved to a separate class that will be responsible for managing the state lifecycle.
+
+- `SwirldStateManager` needs refactoring. The current implementation has two sets of responsibilities:
+  - Keeping track of the references to the latest immutable state and the latest mutable state. This includes creating a copy of the state and updating the references.
+  - Handling Platform-related events:
+    - Handling consensus rounds.
+    - Sealing consensus rounds.
+    - Keeping the `freeze time` parameter up-to-date.
+    - Providing state for signing.
+    - Initializing the state by extracting it from the signed state.
+
+These two sets of responsibilities should be separated. The `swirlds-state-impl` should not have any details related to the platform. However, it should provide all necessary levers for the platform to interact with the state and its lifecycle.
+
+Therefore, `swirlds-state-impl` needs a `StateManager` class that will have the following responsibilities:
```

Comment:
Does it make sense to have a state manager interface in the `swirlds-state-api` module, and one more class that implements the interface in `swirlds-state-impl`?

Reply:
The idea was to have an implementation only, but you're right - we'd better have `StateManager` as an interface and `StateManagerImpl` as an implementation